ads/google/a4a/test/test-google-ads-a4a-config.js

```diff
@@ -272,5 +291,100 @@ describe('a4a_config', () => {
+        href: 'https://cdn.ampproject.org/fnord',
+        pathname: '/fnord',
+        origin: 'https://cdn.ampproject.org',
+        hash: '',
+        search: 'somewhere=over&the=rainbow',
+      },
+      document: {
+        hidden: false,
+        cookie: null,
+        visibilityState: 'visible',
+        addEventListener: function(type, listener) {
+          events[type] = listener;
+        },
+      },
+      crypto: {
+        subtle: true,
+        webkitSubtle: true,
+      },
+      navigator: window.navigator,
+    };
+    events = {};
+    platformFor(win);
+    documentStateFor(win);
+  });
+  afterEach(() => {
+    resetExperimentToggles_();  // Clear saved, page-level experiment state.
+    resetServiceForTesting(win, 'viewer');
+    sandbox.restore();
+  });
 
+  const hashBaseConditions = ['#exp=PARAM',
+    '#p=blarg&exp=PARAM',
+    '#p=blarg&exp=PARAM&s=987',
+    '#p=blarg&exp=zort:123,PARAM,spaz:987&s=987'];
+
+  hashBaseConditions.forEach(hashBase => {
+    it(`should find viewer param when pattern is ${hashBase}`, () => {
+      win.location.hash = hashBase.replace('PARAM', 'a4a:-1');
+      installViewerService(win);
+      // Ensure that internal branches aren't attached, even if the PRNG
+      // would normally trigger them.
+      rand.onFirstCall().returns(-1);
+      const element = document.createElement('div');
+      // Should not register as 'A4A enabled', but should still attach the
+      // control experiment ID.
+      expect(googleAdsIsA4AEnabled(win, element, EXP_ID, EXTERNAL_BRANCHES,
+          INTERNAL_BRANCHES), 'googleAdsIsA4AEnabled').to.be.true;
+      expect(win.document.cookie).to.be.null;
+      expect(rand, 'rand never called').to.not.be.called;
+      expect(isInManualExperiment(element), 'element in manual experiment')
+          .to.be.true;
+      // And it shouldn't be in any *other* experiments.
+      for (const branch in EXTERNAL_BRANCHES) {
+        expect(isInExperiment(element, EXTERNAL_BRANCHES[branch]),
+            'element in ', EXTERNAL_BRANCHES[branch]).to.be.false;
+      }
+      for (const branch in EXTERNAL_BRANCHES) {
+        expect(isInExperiment(element, INTERNAL_BRANCHES[branch]),
+            'element in ', EXTERNAL_BRANCHES[branch]).to.be.false;
+      }
+    });
+
+    it(`hash should trump search; pattern=${hashBase}`, () => {
+      win.location.search = hashBase.replace('PARAM', 'a4a:-1');
+      win.location.hash = hashBase.replace('PARAM', 'a4a:2');
+      installViewerService(win);
+      // Ensure that internal branches aren't attached, even if the PRNG
+      // would normally trigger them.
+      rand.onFirstCall().returns(-1);
+      const element = document.createElement('div');
+      expect(googleAdsIsA4AEnabled(win, element, EXP_ID, EXTERNAL_BRANCHES,
+          INTERNAL_BRANCHES), 'googleAdsIsA4AEnabled').to.be.true;
+      expect(win.document.cookie).to.be.null;
+      expect(rand.called, 'rand called at least once').to.be.false;
```

Comment:
one more left.

I think you can also remove the message, as it will say something similar for you.